string/string.mbt

```diff
@@ -11,3 +11,190 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
+pub fn iter(self : String, f : (Char) -> Unit) {
+  for i = 0; i < self.length(); i = i + 1 {
+    f(self[i])
+  }
+}
+
+test "iter" {
+  let mut i = 0
+  let mut failed = false
+  "12345".iter(
+    fn(elem) { if elem != (i + 1).to_string()[0] { failed = true }; i = i + 1 },
+  )
+  @assertion.assert_false(failed)?
+}
+
+pub fn iteri(self : String, f : (Int, Char) -> Unit) {
+  for i = 0; i < self.length(); i = i + 1 {
+    f(i, self[i])
+  }
+}
+
+test "iteri" {
+  let mut i = 0
+  let mut failed = false
+  "12345".iteri(
+    fn(index, elem) {
+      if index != i || elem != (i + 1).to_string()[0] {
+        failed = true
+      }
+      i = i + 1
+    },
+  )
+  @assertion.assert_false(failed)?
+}
+
+pub fn iter_rev(self : String, f : (Char) -> Unit) {
+  for i = self.length() - 1; i >= 0; i = i - 1 {
+    f(self[i])
+  }
+}
+
+test "iter rev" {
+  let mut i = 6
+  let mut failed = false
+  "12345".iter_rev(
+    fn(elem) { if elem != (i - 1).to_string()[0] { failed = true }; i = i - 1 },
+  )
+  @assertion.assert_false(failed)?
+}
+
+pub fn iteri_rev(self : String, f : (Int, Char) -> Unit) {
+  for i = self.length() - 1; i >= 0; i = i - 1 {
+    f(i, self[i])
+  }
+}
+
+test "iteri rev" {
+  let mut i = 6
+  let mut failed = false
+  "12345".iteri_rev(
+    fn(index, elem) {
+      if index != i - 2 || elem != (i - 1).to_string()[0] {
+        failed = true
+      }
+      i = i - 1
+    },
+  )
+  @assertion.assert_false(failed)?
+}
+
+pub fn map(self : String, f : (Char) -> Char) -> String {
+  if self.length() == 0 {
+    return ""
+  }
+  for i = 0, str = ""; i < self.length(); {
+    continue i + 1,
+      str.push(f(self[i]))
+  } else {
+    str
+  }
+}
+
+test "map" {
+  let str = "12345"
+  let strm = str.map(fn(_elem) { '1' })
+  @assertion.assert_eq(strm, "11111")?
+}
+
+/// Convert string to array.
+pub fn to_array(self : String) -> Array[Char] {
+  if self.length() == 0 {
+    return Array::default()
+  }
+  let arr = Array::make(self.length(), self[0])
+  for i = 1; i < self.length(); i = i + 1 {
+    arr[i] = self[i]
+  }
+  arr
+}
+
+test "to_array" {
+  let str = "12345"
+  let array = str.to_array()
+  @assertion.assert_eq(array, ['1', '2', '3', '4', '5'])?
+}
+
+/// Add a character at the end of the string
+pub fn push(self : String, ch : Char) -> String {
+  self + ch.to_string()
+}
+
+test "push" {
+  let str = push("1234", '5')
+  @assertion.assert_eq(str, "12345")?
+}
+
+/// Reserve a string
+pub fn reserve(self : String) -> String {
```

Comment:
This might turn a valid UTF-8 string into an invalid one

```
test "reverse" {
  let a = "你好"
  let b = a.reserve()
  let c = b.reserve()
  @assertion.assert_eq(a, c)? // fail `"你好" == "¤Ã½Â Â¥Ã¥Â½Â"`
}
```